sys/initrd.c

```diff
@@ -0,0 +1,279 @@
+}
+
+static void read_bytes(char **tape, void *ptr, size_t bytes) {
+  memcpy(ptr, *tape, bytes);
+  *tape += bytes;
+}
+
+static void skip_bytes(char **tape, size_t bytes) {
+  *tape = align(*tape + bytes, 4);
+}
+
+#define MKDEV(major, minor) (((major & 0xff) << 8) | (minor & 0xff))
+
+static bool read_cpio_header(char **tape, cpio_node_t *cpio) {
+  cpio_new_hdr_t hdr;
+
+  read_bytes(tape, &hdr, sizeof(hdr));
+
+  uint16_t c_magic = strntoul(hdr.c_magic, 6, NULL, 8);
+
+  if (c_magic != CPIO_NMAGIC && c_magic != CPIO_NCMAGIC) {
+    log("wrong magic number: %o", c_magic);
+    return false;
+  }
+
+  uint32_t c_ino = strntoul(hdr.c_ino, 8, NULL, 16);
+  uint32_t c_mode = strntoul(hdr.c_mode, 8, NULL, 16);
+  uint32_t c_uid = strntoul(hdr.c_uid, 8, NULL, 16);
+  uint32_t c_gid = strntoul(hdr.c_gid, 8, NULL, 16);
+  uint32_t c_nlink = strntoul(hdr.c_nlink, 8, NULL, 16);
+  uint32_t c_mtime = strntoul(hdr.c_mtime, 8, NULL, 16);
+  uint32_t c_filesize = strntoul(hdr.c_filesize, 8, NULL, 16);
+  uint32_t c_maj = strntoul(hdr.c_maj, 8, NULL, 16);
+  uint32_t c_min = strntoul(hdr.c_min, 8, NULL, 16);
+  uint32_t c_rmaj = strntoul(hdr.c_rmaj, 8, NULL, 16);
+  uint32_t c_rmin = strntoul(hdr.c_rmin, 8, NULL, 16);
+  uint32_t c_namesize = strntoul(hdr.c_namesize, 8, NULL, 16);
+
+  cpio->c_dev = MKDEV(c_maj, c_min);
+  cpio->c_ino = c_ino;
+  cpio->c_mode = c_mode;
+  cpio->c_nlink = c_nlink;
+  cpio->c_uid = c_uid;
+  cpio->c_gid = c_gid;
+  cpio->c_rdev = MKDEV(c_rmaj, c_rmin);
+  cpio->c_size = c_filesize;
+  cpio->c_mtime = c_mtime;
+
+  cpio->c_path = *tape;
+  skip_bytes(tape, c_namesize);
+  cpio->c_data = *tape;
+  skip_bytes(tape, c_filesize);
+
+  return true;
+}
+
+static void read_cpio_archive() {
+  char *tape = (char *)rd_start;
+
+  while (true) {
+    cpio_node_t *node = kmalloc(mp, sizeof(cpio_node_t), M_ZERO);
+    TAILQ_INIT(&node->c_children);
+    if (!read_cpio_header(&tape, node) ||
+        strcmp(node->c_path, CPIO_TRAILER) == 0) {
+      kfree(mp, node);
+      break;
+    }
+    TAILQ_INSERT_TAIL(&initrd_head, node, c_list);
+  }
+}
+
+static bool is_direct_descendant(const char *A, const char *B) {
+  const char *ai = A, *bi = B;
+  /* if B is empty or dot, A is of form entry_name where
+   * entry_name has no "/" character */
+  if (*bi == '\0' || *bi == '.') {
+    if (*bi == '.')
+      bi++;
+    while (*ai && *ai != '/')
+      ai++;
+    if (*ai == '\0')
+      return true;
+    return false;
+  }
+  /* order to check if A is direct descendant of B
+   * it is enough to check if A is of the form "B/entry_name"
+   * where entry_name does not contain "/" character */
+  else {
+    while (*ai && *ai == *bi) {
+      ai++;
+      bi++;
+    }
+    if (*bi != '\0' || *ai != '/')
+      return false;
+    ai++;
+    while (*ai && *ai != '/')
+      ai++;
+    if (*ai == '\0')
+      return true;
+    return false;
+  }
+}
+
+void initrd_construct_c_name(cpio_node_t *cn) {
```

Comment:
Remember to declare function as `static` if you don't want to expose them to other modules in the kernel.